classes/core/PKPHtmlSanitizer.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+/**
+ * @file classes/core/PKPHtmlSanitizer.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class PKPHtmlSanitizer
+ *
+ * @ingroup core
+ *
+ * @brief Wrapper on the top of Symfony's HtmlSanitizer implementation
+ *
+ */
+
+namespace PKP\core;
+
+use DOMDocument;
+use Illuminate\Support\Str;
+use Illuminate\Support\Collection;
+use PKP\config\Config;
+use Symfony\Component\HtmlSanitizer\HtmlSanitizer;
+use Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig;
+
+class PKPHtmlSanitizer
+{
+    /**
+     * Collection of allowed tags to allowed attributes map as key/value[array] structure
+     */
+    protected Collection $allowedTagToAttributeMap;
+
+    /**
+     * Instance of HtmlSanitizerConfig config
+     */
+    protected ?HtmlSanitizerConfig $htmlSanitizerConfig = null;
+
+    /**
+     * Create a new instance
+     */
+    public function __construct(string $allowable)
+    {
+        $this->allowedTagToAttributeMap = $this->generateAllowedTagToAttributeMap(Config::getVar('security', $allowable, null) ?? $allowable);
+    }
+
+    /**
+     * Set a pre defined instance of \Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig
+     */
+    public function setSanitizerConfig(HtmlSanitizerConfig $htmlSanitizerConfig): self
+    {
+        $this->htmlSanitizerConfig = $htmlSanitizerConfig;
+
+        return $this;
+    }
+
+    /**
+     * Get the instance of \Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig
+     */
+    public function getSanitizerConfig(): HtmlSanitizerConfig
+    {
+        if (!$this->htmlSanitizerConfig) {
+            $this->buildSanitizerConfig();
+        }
+
+        return $this->htmlSanitizerConfig;
+    }
+
+    /**
+     * Sanitize the given html string
+     */
+    public function sanitize(string $html): string
+    {   
+        $config = clone $this->getSanitizerConfig();
+
+        $this
+            ->getAllNotAllowedHtmlTags($html)
```

Comment:
We only have two use cases for `PKPHtmlSanitizer` currently, right?

1. When sanitizing a snippet of HTML that has generally been entered by TinyMCE
2. When getting a small piece out of an entire HTML / XML document (as in the two plugins)

Probably 99% of our usage is going to be the first case; using `getAllNotAllowedHtmlTags` is going to incur an entire DOM load for each of these cases, when it's not generally relevant to that use case. And the load could be substantial, as `|strip_unsafe_html` is used in many front-end templates.

I think we might need to give up trying to solve those use cases separately. And in fact they're different enough that I think it's justified; in 1) we're sanitizing user-entered HTML, and in 2) we're extracting part of a JATS XML document.

With use case 2 removed from consideration, is there still justification for the `PKPHtmlSanitizer` class?

Reply:
> going to incur an entire DOM load for each of these cases, when it's not generally relevant to that use case. And the load could be substantial

agreed and better to avoid that . 

> is there still justification for the PKPHtmlSanitizer class

I still think we should have one default sanitization process rather than multiple just because one is more frequently used than other . 

@asmecher I think my previous way to thinking for this implementation was flawed . Basically we need to work with W3C sanitization specification and and let other things handle differently . This newer implementation of `PKPHtmlSanitizer` use the following approach 
1. Build up allowed list
2. cross match with W3C provided list and build up the non allowed list (also include dangerous tags e.g. `script`,  etc)
3. any other tags not in the combined list of allowed and non-allowed list, should be stripped . 

This make it much better and faster to work with and still we can maintain a proper standard specify via W3C sanitization API . 

more about the default configuration at https://wicg.github.io/sanitizer-api/#default-configuration . 

can you review it again ?